api/v1beta1/placementapi_types.go

```diff
@@ -86,6 +86,11 @@ type PlacementAPISpec struct {
 	// actual action pod gets started with sleep infinity
 	Debug PlacementAPIDebug `json:"debug,omitempty"`
 
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// AdvancedMode - allows some usage not fully supported by this operator.
```

Comment:
We need to be more explicit about documenting what is allowed.

Reply:
> We need to be more explicit about documenting what is allowed.

For me, this `advancedMode` terminology is too broad for this reason. We need to assume that at best users might read the documentation, but most likely they will just read the spec. And `advancedMode` doesn't tell me anything about what it does. Is this some kind of Nova advancedMode? Can I find out about it by reading the Nova docs? Does it allow the operator do advanced things? 

I personally would prefer to see something like feature sets, where we enable and disable specific features with clearly defined scoped. `spec.enabledFeatures.databaseMigration`. This to me would be a more intuitive interface, and the expectations of such a setting are fairly obvious from the naming convention.

My other question would be, how common of an interface is this in other Kubernetes primitives or operators? Do we see boolean configurable "modes" elsewhere? Or is our implementation bespoke and therefore not quite as  intuitive to the average Kubernetes user? Keeping in mind that a large selling point for this solution is "You don't need to have TripleO specific knowledge / Heat specific knowledge / Red Hat specific knowledge. You just need to know kubernetes." I think from a design perspective, we should keep that fairly high on the priority list of design considerations.

Reply:
> For me, this advancedMode terminology is too broad for this reason. We need to assume that at best users might read the documentation, but most likely they will just read the spec. And advancedMode doesn't tell me anything about what it does. Is this some kind of Placement advancedMode? Can I find out about it by reading the Placement docs? Does it allow the operator do advanced things?

It is the latter. :) 

On this point I'm OK to group the operator specific configuration if there are many (like advancedMode, preserveJobs) into a separate struct to express that they are distinct from the configuration that is for the deployed OpenStack service (like customeServiceConfig). 

> I personally would prefer to see something like feature sets, where we enable and disable specific features with clearly defined scopes. eg. spec.enabledFeatures.databaseMigration. This to me would be a more intuitive interface, and the expectations of such a setting are fairly obvious from the naming convention.

My problem with this that we might end up with a long list of flags if we start adding a configurable for each different "advanced feature". Like the next would be allowing changing the memcached instance under a flag. 

> My other question would be, how common of an interface is this in other Kubernetes primitives or operators? Do we see boolean configurable "modes" elsewhere? Or is our implementation bespoke and therefore not quite as intuitive to the average Kubernetes user? 

I'm not sure how we define 'boolean configurable "modes"' but it was fairly easy to find examples of boolean flags in Spec structs that's value affecting the behavior of the operator itself instead of affecting the behavior of the running service deployed by the operator:
* https://github.com/openstack-k8s-operators/rabbitmq-cluster-operator/blob/f8d9ac5a263c5d9c82dfeef8170eacf16c1cc5e1/api/v1beta1/rabbitmqcluster_types.go#L76-L80 
* JobSpec.Suspended https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#job-v1-batch 

> Keeping in mind that a large selling point for this solution is "You don't need to have TripleO specific knowledge / Heat specific knowledge / Red Hat specific knowledge. You just need to know kubernetes." I think from a design perspective, we should keep that fairly high on the priority list of considerations.

If we allow everything that k8s allows and ignoring what OpenStack supports from the set of operations k8s allows, then I think we will end up in too many dangerous possibilities. I think one of the value we can bring to the customers is codified knowledge about deploying OpenStack, encoded into the implementation of the operators. If we know that changing a DB server under a running OpenStack deployment is dangerous for the integrity of the deployment, then we should tell that to our users somehow. I also agree that we should not assume that they read the doc. So I would codify that knowledge into the operator.

Reply:
Hey @gibizer thanks for the response.

I think in both of those examples it's more clear what the setting is though. It's not a broadly scoped `advancedMode` is more my point there. I think we want to avoid a situation where people just think, "Advanced mode! I want to do advanced things! I'll set this to true."

The alternative of `databaseMigration` makes you think a little more. "Do I want to be doing a database migration? I don't think so. I'll leave it false."

It's not that we want to allow k8s things. We just want to make sure the interface is intuitive for people familiar with k8s. I'm not too sure there are examples of other operators implementing an `advancedMode` that allows an arbitrary list of operations is all. In this case, we're talking pretty specifically about a k8s operator feature where we would implement job logic to migrate databases around. The OpenStack service doesn't necessarily need to be aware or care, it just needs a new address for the DB service and a restart. So I think in this case, we're implementing an interface on a k8s CRD which seems more specific to a k8s implementation detail rather than OpenStack implementation detail.

But yeah, I understand your concern around an all encompassing list of possible things that can be done. I would guess at a max of 5 per service though which doesn't seem too bad.

In any case, I'm just a single voice. If we have a consensus on the approach proposed here, I'm happy to work with it. But I thought it would be worth at least raising my concern and suggestion for consideration before we merge it.

:)

Thanks again for the perspective and happy Monday.